src/allocation_map.rs

```diff
@@ -0,0 +1,272 @@
+            let elem = &self.v[candidate];
+            if offset < elem.range.start {
+                // We are too far right (offset is further left).
+                debug_assert!(candidate < right); // we are making progress
+                right = candidate;
+            } else if offset >= elem.range.end() {
+                // We are too far left (offset is further right).
+                debug_assert!(candidate >= left); // we are making progress
+                left = candidate + 1;
+            } else {
+                // This is it!
+                return Ok(candidate);
+            }
+        }
+    }
+
+    /// Determines whether a given access on `range` overlaps with
+    /// an existing allocation
+    pub fn access_type(&self, range: AllocRange) -> AccessType {
+        match self.find_offset(range.start) {
+            Ok(index) => {
+                // Start of the range belongs to an existing object, now let's check the overlapping situation
+                let elem = &self.v[index];
+                // FIXME: derive Eq for AllocRange in rustc
+                if elem.range.start == range.start && elem.range.size == range.size {
+                    // Happy case: perfectly overlapping access
+                    AccessType::PerfectlyOverlapping(index)
+                } else {
+                    // FIXME: add a last() method to AllocRange that returns the last inclusive offset (end() is exclusive)
+                    let end_index = match self.find_offset(range.end() - Size::from_bytes(1)) {
+                        // If the end lands in an existing object, add one to get the exclusive index
+                        Ok(inclusive) => inclusive + 1,
+                        Err(exclusive) => exclusive,
+                    };
+
+                    AccessType::ImperfectlyOverlapping(index..end_index)
+                }
+            }
+            Err(index) => {
+                // Start of the range doesn't belong to an existing object
+                match self.find_offset(range.end() - Size::from_bytes(1)) {
+                    // Neither does the end
+                    Err(end_index) =>
+                        if index == end_index {
+                            // There's nothing between the start and the end, so the range thing is empty
+                            AccessType::Empty(index)
+                        } else {
+                            // Otherwise we have entirely covered an existing object
+                            AccessType::ImperfectlyOverlapping(index..end_index)
+                        },
+                    // Otherwise at least part of it overlaps with something else
+                    Ok(end_index) => AccessType::ImperfectlyOverlapping(index..end_index + 1),
+                }
+            }
+        }
+    }
+
+    /// Inserts an object and its occupied range at given position
+    pub fn insert(&mut self, index: Position, range: AllocRange, data: T) {
```

Comment:
It seems odd to ask the user to give the position, rather than determining that automatically?

Reply:
Ideally we really want `insert`, `remove`, `Index` and `IndexMut` all take in `AllocRange`, but then they all have to do a binary search on each call. In `StoreBufferAlloc::get_store_buffer(_mut)` it always calls `access_type()` first, letting that return the index from binary search and then reusing them avoids repeating the search.

I may be prematurely optimising here. I'm more than happy to make all methods determining the actual index on its own

Reply:
That makes sense; the API is designed for a single consumer after all. Just a brief comment explaining this would be good.

Also, for extra type safety, can you make `Position` a proper newtype with a private field?

Reply:
I tried to make it a newtype but then I can't iterate over `Range<Position>` because it needs to implement `Step` which is non-derivable and [looks like quite a lot of work](https://doc.rust-lang.org/std/iter/trait.Step.html)...

Reply:
You could probably just call the existing `usize as Step` impl, but yeah it's probably not worth it at that point.